engine/src/flutter/lib/web_ui/lib/src/engine/canvaskit/surface.dart

```diff
@@ -2,557 +2,342 @@
-        _updateLogicalHtmlCanvasSize();
-      }
+    _skSurface?.dispose();
+    _skSurface = _createSoftwareSkSurface();
+    if (_skSurface == null) {
+      throw Exception('Failed to initialize CanvasKit SkSurface.');
     }
+  }
 
-    // When the browser tab using WebGL goes dormant the browser and/or OS may
-    // decide to clear GPU resources to let other tabs/programs use the GPU.
-    // When this happens, the browser sends the "webglcontextlost" event as a
-    // notification. When we receive this notification we force a new context.
-    //
-    // See also: https://www.khronos.org/webgl/wiki/HandlingContextLost
-    _cachedContextRestoredListener = createDomEventListener(_contextRestoredListener);
-    _cachedContextLostListener = createDomEventListener(_contextLostListener);
-    htmlCanvas.addEventListener('webglcontextlost', _cachedContextLostListener, false.toJS);
-    htmlCanvas.addEventListener('webglcontextrestored', _cachedContextRestoredListener, false.toJS);
-    _forceNewContext = false;
-    _contextLost = false;
-
-    if (webGLVersion != -1 && !configuration.canvasKitForceCpuOnly) {
-      int glContext = 0;
-      final SkWebGLContextOptions options = SkWebGLContextOptions(
-        // Default to no anti-aliasing. Paint commands can be explicitly
-        // anti-aliased by setting their `Paint` object's `antialias` property.
-        antialias: _kUsingMSAA ? 1 : 0,
-        majorVersion: webGLVersion.toDouble(),
-      );
-      if (useOffscreenCanvas) {
-        glContext = canvasKit.GetOffscreenWebGLContext(_offscreenCanvas!, options).toInt();
-      } else {
-        glContext = canvasKit.GetWebGLContext(_canvasElement!, options).toInt();
-      }
+  /// Creates an SkSurface for software rendering. This is used when WebGl is not
+  /// supported or when it fails to initialize.
+  SkSurface _createSoftwareSkSurface();
 
-      _glContext = glContext;
-
-      if (_glContext != 0) {
-        _grContext = canvasKit.MakeGrContext(glContext.toDouble());
-        if (_grContext == null) {
-          // TODO(harryterkelsen): Make this error message more descriptive by
-          // reporting the number of currently live Surfaces, https://github.com/flutter/flutter/issues/162868.
-          throw CanvasKitError(
-            'Failed to initialize CanvasKit. '
-            'CanvasKit.MakeGrContext returned null.',
-          );
-        }
-        if (_sampleCount == -1 || _stencilBits == -1) {
-          _initWebglParams();
-        }
-        // Set the cache byte limit for this grContext, if not specified it will
-        // use CanvasKit's default.
-        _syncCacheBytes();
-      }
+  double _currentDevicePixelRatio = -1;
+
+  @override
+  Future<void> setSize(BitmapSize size) async {
+    final double devicePixelRatio = EngineFlutterDisplay.instance.devicePixelRatio;
+    if (_skSurface != null &&
+        _skSurface!.width() == size.width &&
+        _skSurface!.height() == size.height &&
```

Comment:
Can be done without calling to CanvasKit:

```suggestion
        size == _currentSize &&
```